File: data_structure/sort.c
#define SWAP(x, y, t) ( (t) = (x), (x) = (y), (y) = (t) )

// 선택정렬
void selection_sort(int list[], int n)
{
    int i, j, least, temp;
    for (i = 0; i < n - 1; i++) {
        least = i;
        for (j = i + 1; j < n; j++) {
            if (list[j] < list[least]) least = j;
        SWAP(list[i], list[least], temp);
        }
    }
}

// 삽입정렬
void insertion_sort(int list[], int n)
{
    int i, j, key;
    for (i = 1; i < n; i++) {
        key = list[i];
        for (j = i - 1; j >= 0 && list[j] > key; j--)
            list[j + 1] = list[j];
        list[j + 1] = key;
    }
}

// 버블정렬
void bubble_sort(int list[], int n)
{
    int i, j, temp;
    for (i = n - 1; i > 0; i--) {
        for (j = 0; j < i; j++) {
            if (list[j] > list[j + 1])
                SWAP(list[j], list[j + 1], temp);
        }
    }
}

// 퀵정렬
#define MAX_SIZE 10
int list[MAX_SIZE];
int n;

int partition(int list[], int left, int right)
{
    int pivot, temp;
    int low, high;
    // 첫번쨰 요소는 비교에서 제외시키니까 + 1 안함
    low = left;
    // 첫번째 요소 포함시키려면 + 1 해야됨
    high = right + 1;
    pivot = list[left];
    do {
        do
        {
            low++;
        } while (list[low] < pivot);
        do
        {
            high--;
        } while (list[high] > pivot);
        if (low < high) SWAP(list[low], list[high], temp);
    } while (low < high);

    SWAP(list[left], list[high], temp);
    return high;
}

void quick_sort(int list[], int left, int right)
{
    if (left < right)
    {
        int q = partition(list, left, right);
        quick_sort(list, left, q - 1);
        quick_sort(list, q + 1, right);
    }
}

// 병합 정렬
int sorted[MAX_SIZE]; /*병합 정렬은 추가공간 필요*/

/* i는 정렬된 왼쪽 리스트에 대한 인덱스
    j는 정렬된 오른쪽 리스트에 대한 인덱스
    k는 정렬될 리스트에 대한 인덱스
*/
void merge(int list[], int left, int mid, int right)
{
    int i, j, k, l;
    i = left; j = mid + 1, k = left;

    /* 분할 정렬된 리스트의 병합*/
    while (i <= mid && j <= right) {
        if (list[i] <= list[j])
            sorted[k++] = list[i++];
        else
            sorted[k++] = list[j++];
    }
    /* 남아있는 레코드 일괄 복사 */
    if (i > mid)
        for (l = j; l <= right; l++)
            sorted[k++] = list[l];
    else
        for (l = i; l <= mid; l++)
            sorted[k++] = list[l];
    /* 임시 배열에서 list로 재복사 */
    for (l = left; l <= right; l++)
        list[l] = sorted[l];
}

void merge_sort(int list[], int left, int right)
{
    int mid;
    if (left < right) {
        mid = (left + right) / 2;
        merge_sort(list, left, mid);
        merge_sort(list, mid + 1, right);
        merge(list, left, mid, right);
    }
}

File: data_structure/sort.py
target = [3, 5, 1, 2, 7, 8, 4, 6, 9, 10]

def selection_sort(target):
    l = len(target)
    for i in range(0, l - 1):
        min = i
        for j in range(i + 1, l):
            if target[j] < target[min]:
                min = j
        target[i], target[min] = target[min], target[i]

selection_sort(target)
target = [3, 5, 1, 2, 7, 8, 4, 6, 9, 10]

def insertion_sort(target):
    l = len(target)
    for i in range(1, l):
        temp = i
        for j in range(i - 1, -1, -1):
            if target[temp] < target[j]:
                target[temp], target[j] = target[j], target[temp]
                temp = j
            else:
                break
insertion_sort(target)

target = [3, 5, 1, 2, 7, 8, 4, 6, 9, 10]
def bubble_sort(target):
    l = len(target)
    for i in range(l - 1, 0, -1):
        for j in range(1, i + 1):
            if target[j - 1] > target[j]:
                target[j], target[j - 1] = target[j - 1], target[j]
bubble_sort(target)

target = [3, 5, 1, 2, 7, 8, 4, 6, 9, 10]

def quick_sort(left, right):
    if right - left < 1:
        return

    pivot = target[left]

    lptr, rptr = left + 1, right

    while lptr <= rptr:
        if pivot < target[lptr] and target[rptr] < pivot:
            target[lptr], target[rptr] = target[rptr], target[lptr]
            lptr, rptr = left + 1, right

        if target[lptr] <= pivot:
            lptr += 1

        if pivot <= target[rptr]:
            rptr -= 1

    target[left], target[rptr] = target[rptr], target[left]

    quick_sort(left, rptr - 1)
    quick_sort(lptr, right)

print('before:\t', target)
quick_sort(0, len(target) - 1)
print('after:\t', target)

target = [3, 5, 1, 2, 7, 8, 4, 6, 9, 10]
sorted_list = [0 for _ in range(len(target))]


def merge(left, mid, right):
    i, j, k = left, mid + 1, left

    while i <= mid and j <= right:
        if target[i] <= target[j]:
            sorted_list[k] = target[i]
            i += 1
        else:
            sorted_list[k] = target[j]
            j += 1
        k += 1

    if mid < i:
        for t in range(j, right + 1):
            sorted_list[k] = target[t]
            k += 1
    else:
        for t in range(i, mid + 1):
            sorted_list[k] = target[t]
            k += 1

    for idx in range(left, right + 1):
        target[idx] = sorted_list[idx]


def merge_sort(left, right):
    if left != right:
        mid = (left + right) // 2
        merge_sort(left, mid)
        merge_sort(mid + 1, right)
        merge(left, mid, right)


print('before:\t', target)
merge_sort(0, len(target) - 1)
print('after:\t', target)

import heapq

target = [3, 5, 1, 2, 7, 8, 4, 6, 9, 10]


def heap_sort():
    heapq.heapify(target)

    answer = []
    while target:
        answer.append(heapq.heappop(target))

    return answer


print('before:\t', target)
target = heap_sort()
print('after:\t', target)